traits/observers/_named_trait_observer.py

```diff
@@ -43,12 +60,153 @@ def notify(self):
         """
         return self._notify
 
+    @property
+    def optional(self):
```

Comment:
Thought: should this class be a subclass of named tuple`?  This would get you a lot of the infrastructure for free (you wouldn't need properties and shadow attributes, `__hash__` will just work, and `__eq__` may also just work or need minimal wrapping.